### PR TITLE
fix managed pointer macrons

### PR DESCRIPTION
**Pull request by dtaller**
Some fixes for HIP.

I did not test this yet. I am just creating the pull request before I leave for the day so that tests run. I'll revisit tomorrow.

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
Thanks @dtaller - it might be good to implement this as a new `CHAI_DEVICE_COMPILATION` macro. There are a lot of places that now have the `cudacc || hipcc` check, and as we add more and more backends those checks will get more complicated.

Thoughts @adayton1 ?

**Comment by dtaller** [comment]
> Thanks @dtaller - it might be good to implement this as a new `CHAI_DEVICE_COMPILATION` macro. There are a lot of places that now have the `cudacc || hipcc` check, and as we add more and more backends those checks will get more complicated.
> 
> Thoughts @adayton1 ?

Yes, that is actually what I did for CARE, where I have

#if defined(__CUDACC__) || defined(__HIPCC__)
#define __GPUCC__
#endif

I'll make that change soon. 

I just went through and added gpuFree as a way to do either hipFree or cudaFree (and similar for amlloc and memcpy). Hopefully that is also ok

**Comment by davidbeckingsale** [comment]
Yes, that sounds good - thanks!

**Comment by dtaller** [comment]
Hi @davidbeckingsale and @adayton1 . I think that this is ready for review. 

It compiles now for HIP. When I try to run tests in conjunction with CARE I do have some issues, but I think those are CARE related, not CHAI issues that should be fixed in this pull request. 

Please tell me what you think

File: src/chai/managed_ptr.hpp [hunk 1]
```diff
@@ -272,7 +272,7 @@ namespace chai {
             m_gpu_pointer(other.m_gpu_pointer),
             m_pointer_record(other.m_pointer_record)
          {
-#ifndef __CUDA_ARCH__
+#if !defined(__CUDA_ARCH__) && !defined(__HIP_DEVICE_COMPILE__)
```
Comment (davidbeckingsale) [inline src/chai/managed_ptr.hpp:275]: Can you come up with another more generic macro for this case?

**Comment by davidbeckingsale** [comment]
Looks good @dtaller - just one quick question ☝️

File: src/chai/managed_ptr.hpp (hunk 1 above)
Comment (dtaller) [reply src/chai/managed_ptr.hpp]: Done. I added another macro. 

I tested it by inserting the following code block into the code:

#if !defined(__CUDA_ARCH__) && !defined(__HIP_DEVICE_COMPILE__)
#warning compiling host previous macro
#else
#warning compiling device previous macro
#endif

#if !defined(__DEVICE_COMPILE__)
#warning compiling host danny macro
#else
#warning compiling device  danny macro
#endif

And it came out consistent:

/usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/tpl/chai/src/chai/managed_ptr.hpp:282:2: warning: compiling device previous macro [-W#warnings]
#warning compiling device previous macro
 ^
/usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/tpl/chai/src/chai/managed_ptr.hpp:288:2: warning: compiling device  danny macro [-W#warnings]
#warning compiling device  danny macro
 ^
2 warnings generated when compiling for gfx900.
In file included from /usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/src/care/RAJAPlugin.cpp:12:
In file included from /usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/src/care/util.h:12:
In file included from /usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/src/care/care.h:267:
In file included from /usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/src/care/PointerTypes.h:30:
/usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/tpl/chai/src/chai/managed_ptr.hpp:280:2: warning: compiling host previous macro [-W#warnings]
#warning compiling host previous macro
 ^
/usr/workspace/taller1/CARE_TESTING_CORONA_2/CARE/tpl/chai/src/chai/managed_ptr.hpp:286:2: warning: compiling host danny macro [-W#warnings]
#warning compiling host danny macro

**Comment by dtaller** [comment]
> Looks good @dtaller - just one quick question ☝️

Done. Once the checks finish, should be ready

File: src/chai/ChaiMacros.hpp [hunk 2]
```diff
@@ -11,26 +11,59 @@
...
+#define gpuMemcpyHostToHost 0
+#define gpuMemcpyHostToDevice 1
+#define gpuMemcpyDeviceToHost 2
+#define gpuMemcpyDeviceToDevice 3
+#define gpuMemcpyDefault 4
+
+#endif
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+#define __DEVICE_COMPILE__
```
Comment (davidbeckingsale) [inline src/chai/ChaiMacros.hpp:61]: I think this should be `CHAI_DEVICE_COMPILE`

File: src/chai/ChaiMacros.hpp [hunk 3]
```diff
@@ -11,26 +11,59 @@
...
+#endif
+
+// shorthand for GPU Compilation. Must go after hip/hip_runtime.h is included so that HIPCC is defined
+#if defined(__CUDACC__) || defined(__HIPCC__)
+#define __GPUCC__
```
Comment (davidbeckingsale) [inline src/chai/ChaiMacros.hpp:66]: Can you change this to `CHAI_GPUCC` please.

Comment (dtaller) [reply src/chai/ChaiMacros.hpp]: Changed.

File: src/chai/ChaiMacros.hpp (hunk 2 above)
Comment (dtaller) [reply src/chai/ChaiMacros.hpp]: Changed.

**Comment by dtaller** [comment]
> Please updated the macro names - I'm worried about the generic names ending up causing conflicts with other libraries.

I changed the names of the macros. Tests running now

**Pull request merged**